### DISCUSSION TITLE
Can input keyboard characters into the textbox in linux with drm?

### DISCUSSION THREAD

**Discussion by Coloryr** [AvaloniaUI/Avalonia#14971] [Q&A] [answered]
can be redirect char to a textbox?

**Comment by thevortexcloud** [comment]
You can if you make a soft keyboard. It's fairly simple to do if you are only dealing with English/Latin characters. You just need a couple rows of buttons and a case shift button that changes which character gets appended to the text box text. EG here is the XAML for one I wrote awhile back while messing around with DRM and Avalonia which is probably enough to give you an idea as to how to do it:

```xaml
<Border BorderThickness="1" BorderBrush="Black" Background="LightGray">
        <StackPanel Orientation="Vertical" HorizontalAlignment="Stretch" VerticalAlignment="Stretch" Focusable="False">
            <StackPanel Orientation="Horizontal">
                <views:VirtualKeyboardKey LowerKey='`' UpperKey="~" />
                <views:VirtualKeyboardKey LowerKey='1' UpperKey="!" />
                <views:VirtualKeyboardKey LowerKey='2' UpperKey="@" />
                <views:VirtualKeyboardKey LowerKey='3' UpperKey="#" />
                <views:VirtualKeyboardKey LowerKey='4' UpperKey="$" />
                <views:VirtualKeyboardKey LowerKey='5' UpperKey="%" />
                <views:VirtualKeyboardKey LowerKey='6' UpperKey="^" />
                <views:VirtualKeyboardKey LowerKey='7' UpperKey="&amp;" />
                <views:VirtualKeyboardKey LowerKey='8' UpperKey="*" />
                <views:VirtualKeyboardKey LowerKey='9' UpperKey="(" />
                <views:VirtualKeyboardKey LowerKey='0' UpperKey=")" />
                <views:VirtualKeyboardKey LowerKey='-' UpperKey="_" />
                <views:VirtualKeyboardKey LowerKey='=' UpperKey="+" />
                <views:VirtualKeyboardKey Mode="Special" SpecialKeyType="Backspace" Content="Backspace" />
            </StackPanel>
            <StackPanel Orientation="Horizontal" Margin="20,0,0,0">
                <views:VirtualKeyboardKey LowerKey='q' UpperKey="Q" />
                <views:VirtualKeyboardKey LowerKey='w' UpperKey="W" />
                <views:VirtualKeyboardKey LowerKey='e' UpperKey="E" />
                <views:VirtualKeyboardKey LowerKey='r' UpperKey="R" />
                <views:VirtualKeyboardKey LowerKey='t' UpperKey="T" />
                <views:VirtualKeyboardKey LowerKey='y' UpperKey="Y" />
                <views:VirtualKeyboardKey LowerKey='u' UpperKey="U" />
                <views:VirtualKeyboardKey LowerKey='i' UpperKey="I" />
                <views:VirtualKeyboardKey LowerKey='o' UpperKey="O" />
                <views:VirtualKeyboardKey LowerKey='p' UpperKey="P" />
                <views:VirtualKeyboardKey LowerKey='[' UpperKey="{}{" />
                <views:VirtualKeyboardKey LowerKey=']' UpperKey="{}}" />
                <views:VirtualKeyboardKey LowerKey='\' UpperKey="|" />
            </StackPanel>
            <StackPanel Orientation="Horizontal" Margin="40,0,0,0">
                <views:VirtualKeyboardKey LowerKey='a' UpperKey="A" />
                <views:VirtualKeyboardKey LowerKey='s' UpperKey="S" />
                <views:VirtualKeyboardKey LowerKey='d' UpperKey="D" />
                <views:VirtualKeyboardKey LowerKey='f' UpperKey="F" />
                <views:VirtualKeyboardKey LowerKey='g' UpperKey="G" />
                <views:VirtualKeyboardKey LowerKey='h' UpperKey="H" />
                <views:VirtualKeyboardKey LowerKey='j' UpperKey="J" />
                <views:VirtualKeyboardKey LowerKey='k' UpperKey="K" />
                <views:VirtualKeyboardKey LowerKey='l' UpperKey="L" />
                <views:VirtualKeyboardKey LowerKey=';' UpperKey=":" />
                <views:VirtualKeyboardKey LowerKey="'" UpperKey="&quot;" />
            </StackPanel>
            <StackPanel Orientation="Horizontal" Margin="60,0,0,0">
                <views:VirtualKeyboardKey Content="Shift" Width="120" SpecialKeyType="Shift" Mode="Special" />
                <views:VirtualKeyboardKey LowerKey='z' UpperKey="Z" />
                <views:VirtualKeyboardKey LowerKey='x' UpperKey="X" />
                <views:VirtualKeyboardKey LowerKey='c' UpperKey="C" />
                <views:VirtualKeyboardKey LowerKey='v' UpperKey="V" />
                <views:VirtualKeyboardKey LowerKey='b' UpperKey="B" />
                <views:VirtualKeyboardKey LowerKey='n' UpperKey="N" />
                <views:VirtualKeyboardKey LowerKey='m' UpperKey="M" />
                <views:VirtualKeyboardKey LowerKey=',' UpperKey="&lt;" />
                <views:VirtualKeyboardKey LowerKey='.' UpperKey="&gt;" />
                <views:VirtualKeyboardKey LowerKey='/' UpperKey="?" />
            </StackPanel>
            <StackPanel Orientation="Horizontal" HorizontalAlignment="Center">
                <views:VirtualKeyboardKey Mode="Special" SpecialKeyType="Space" Width="700" />
            </StackPanel>
        </StackPanel>
    </Border>
```

The advantage of doing it this way is you can easily define new keyboard layouts by just declaring different XAML files. 

If you need to deal with non latin languages you will probably need to come up with a different way.

**Comment by rabbitism** [comment]
We had a virtual numpad implementation in Ursa. But only numpad, because it's almost impossible to  deal with ime in avalonia.

**Reply by Coloryr** [reply]
Thanks

**Reply by thevortexcloud** [reply]
It theoretically can be done, it's just very complicated and a lot of work. I would argue there is a case for Avalonia shipping with a soft keyboard. But that's a very big project so I doubt it will happen any time soon. Avalonia currently handles IME by passing it off to the OS, which on Linux I think is being handled by X11, which obviously wont work with DRM. 

You could theoretically just use X11 instead of DRM with your app running full screen. There are a lot of window managers and what not that could be customised to make sure your app is the only thing visible.

**Reply by Gillibald** [reply]
IME is always something the OS needs to provide

**Reply by thevortexcloud** [reply]
Right, but that's not possible on Linux with DRM as far as I am aware. All the existing implementations require some sort of GUI toolkit/desktop environment. In any case, I am not expecting Avalonia to add it's own IME implementation.

**Comment by kyuranger** [comment]
![1](https://github.com/AvaloniaUI/Avalonia/assets/30720036/bc899029-a0c4-49cd-83dc-b49ee7f517d7)
This is a VirtualKeyboard I wrote myself, which also supports pinyin (but I haven't completely completed pinyin input yet). Although I'm demonstrating this on Windows, it will work on Linux as well.

BTW, I am making this virtual keyboard to complete an embedded device project (Linux DRM).

My only regret is that I want to input Chinese by handwriting, but I don't know how to do it (no clue at all). I thought about doing it by handwriting a graphic and then doing OCR through PaddleSharp. But I don’t even know how to use Avalonia handwriting graphics (I once asked this question in the community, but no one answered), let alone whether it is possible to use PaddleSharp.

**Reply by thevortexcloud** [reply]
> . But I don’t even know how to use Avalonia handwriting graphics (I once asked this question in the community, but no one answered), let

I _imagine_ you can just export the control as a bitmap and then feed it into the OCR engine. However most OCR is designed for fonts, not hand writing.

**Reply by kyuranger** [reply]
But I don’t even know how to implement handwriting drawing, and exporting pictures is another thing.

**Reply by rabbitism** [reply]
![IMG20231124180829.jpg](https://github.com/AvaloniaUI/Avalonia/assets/14807942/a023d6f6-ce3c-4cc8-8b74-979631f16719)

hmm, i tried to implement one long time ago.

**Reply by kyuranger** [reply]
That’s awesome. Can you tell me how you achieved it?

**Reply by kyuranger** [reply]
It suddenly occurred to me that I can achieve recognition through ml.net. My impression is that ml.net can use enough samples to identify whether the animal in the picture is a chicken, a duck, a cow, or something else. This function can be used to achieve this. But how to find enough handwritten samples is also a big problem.

**Reply by thevortexcloud** [reply]
>  But how to find enough handwritten samples is also a big problem.

You are right you can use image recognition for this. However ML.NET's built in classifier is not really designed for this as far as I know. You are better off just using some existing library or making your own using one of the [C# OpenCV bindings](https://github.com/shimat/opencvsharp). But that can be fairly complicated. However most machine learning stuff these days is built for Python. As such you could make a simple networked API in Python and just have your .NET app talk to it over the network. EG This one seems fairly popular: https://github.com/githubharald/SimpleHTR

Alternatively use one of the many existing cloud products such as Google's OCR.

https://cloud.google.com/document-ai

> But how to find enough handwritten samples is also a big problem.

Indeed. The way most big companies solve this is by hiring a bunch of cheap workers/contractors in other countries and then pay them to tag data for them. Almost everything about making an AI model is, at least at first, very labour intensive. To get good recognition you would need thousands of labelled samples from a lot of different people.

**Reply by kyuranger** [reply]
Unfortunately, my embedded device is for factory use and simply runs offline.

The main problem now is still how to implement handwriting drawing.

As for Avalonia handwritten drawing, someone here seems to be able to achieve it:
https://github.com/AvaloniaUI/Avalonia/pull/7412

But I don't know exactly how it is implemented, and it seems to only support the Windows platform.

In addition, I saw that this control can be implemented in other topics, but I am also not sure how to use it:
https://github.com/AvaloniaUI/Avalonia/blob/master/samples/ControlCatalog/Pages/PointerCanvas.cs

**Reply by kyuranger** [reply]
Finally, I made a virtual keyboard (on screen keyboard) that can input English, Chinese (pinyin), symbols, and numbers, but I did not follow the official Avalonia interface (https://github.com/AvaloniaUI/Avalonia/wiki /IME), the performance may not be so ideal.

![1](https://github.com/AvaloniaUI/Avalonia/assets/30720036/b5e21e7e-b1dd-47ca-97c2-3e2c9942d065)

**Comment by Coloryr** [comment]
I am creating a virtual keyboard and connecting it to [RIME](https://github.com/rime/librime)

**Reply by thevortexcloud** [reply]
Very nice. I am not sure what Avalonia's stance is on it, but I suspect a lot of people would appreciate if even just the bindings were added to Avalonia so people can implement their own keyboards.

**Reply by kyuranger** [reply]
BTW, Avalonia has a library that you can implement ( https://github.com/AvaloniaUI/Avalonia/wiki/IME ).It's just that I don't know exactly how to use it.

**Reply by thevortexcloud** [reply]
It looks like it's just the abstraction layer? In which case it still needs an integration with a system IME provider..

**Reply by Coloryr** [reply]
You can send a `Key Event` send to TopLevel?

**Comment by Coloryr** [comment, marked as answer]
Is work now https://github.com/Coloryr/AvaloniaKeyboard, but need https://github.com/Coloryr/librime

**Reply by kyuranger** [reply]
I tried your keyboard and found that the RimeSharp project was missing and could not be compiled.

**Reply by kyuranger** [reply]
Oh, I suddenly discovered that I need to reference another project: https://github.com/Coloryr/RimeSharp/tree/51cdbcf54d19d55444b37e02fb6be9bad9ad7f5c. After I tried to compile this project, I found that other projects also need to be referenced. I couldn't quite figure out how to run it correctly, so I gave up. I am so sorry.

**Reply by Coloryr** [reply]
I'm use submodule, you can try
```
git submodule add https://github.com/Coloryr/AvaloniaKeyboard.git
git submodule update --init --recursive
```
And add project `AvaloniaKeyboard` and `RimeSharp`
then build once
Download the [Action](https://github.com/Coloryr/librime/actions/runs/8323259120) zip, unzip it and copy `rime.dll` to your build dir
last step, copy this [files](https://github.com/Coloryr/librime/tree/master/data/minimal) to `{your build dir}/data`

**Reply by Coloryr** [reply]
I'm going to write a guide now
